### ISSUE TITLE
Recursive call causes runtime error

### ISSUE THREAD

**Issue by AZon8** [elm-lang/elm-compiler#1282]
The following code causes a runtime error. 

        import Graphics.Element exposing (..)
        import Dict exposing (..)

        main =  let 
            dict = update (ActOnSecondLevel "somekey" "secondkey" (\a -> a ++ "b")) Dict.empty
          in show dict

        type alias FirstLevelKey = String
        type alias SecondLevelKey = String
        type alias SomeData = String
        type alias FirstLevelDict = Dict FirstLevelKey SecondLevelDict
        type alias SecondLevelDict = Dict SecondLevelKey SomeData
        type Action =
            ActOnFirstLevel FirstLevelKey (SecondLevelDict -> SecondLevelDict)
            |  ActOnSecondLevel FirstLevelKey SecondLevelKey (SomeData -> SomeData)

        update : Action -> FirstLevelDict -> FirstLevelDict 
        update act model =
            case act of
                ActOnFirstLevel  key1 updater ->  
                    Dict.insert key1 (updater (Maybe.withDefault Dict.empty (Dict.get key1 model))) model
                ActOnSecondLevel  key1 key2 updater -> 
                    update (ActOnFirstLevel key1 (\p ->  Dict.insert key2 (updater (Maybe.withDefault "DefaultData" (Dict.get key2 p))) p ) ) model

**Comment by jvoigtlaender** [comment]
Are you sure that this *should* terminate normally? Elm does not prevent you from writing nonterminating recursions. For example, this program:
```elm
import Graphics.Element

main = Graphics.Element.show (loop "a")

loop = loop
```
also leads to a runtime error.

Is your program of this kind, or are your recursive calls legitimate and you would be entitled to get something different from runtime error?

**Comment by AZon8** [comment]
Thanks for the quick reply. 
I'm not at all familiar with the compiler but. 

It doesn't do much recursion. 
The expected result was :

     Dict.fromList [("somekey",Dict.fromList [("secondkey","DefaultDatab")])]

Which i actually do get when for instance i write it like this and 'update' then calls 'update2' ( the same function but in a different scope) 

    update : Action -> FirstLevelDict -> FirstLevelDict 
    update act model =
        case act of
            ActOnFirstLevel  key1 updater ->  
                Dict.insert key1 (updater (Maybe.withDefault Dict.empty (Dict.get key1 model))) model
            ActOnSecondLevel  key1 key2 updater -> 
                **update2** (ActOnFirstLevel key1 (\p ->  Dict.insert key2 (updater (Maybe.withDefault "DefaultData" (Dict.get key2 p))) p ) ) model


    update2 : Action -> FirstLevelDict -> FirstLevelDict 
    update2 act model =
        case act of
            ActOnFirstLevel  key1 updater ->  
                Dict.insert key1 (updater (Maybe.withDefault Dict.empty (Dict.get key1 model))) model
            ActOnSecondLevel  key1 key2 updater -> 
                update2 (ActOnFirstLevel key1 (\p ->  Dict.insert key2 (updater (Maybe.withDefault "DefaultData" (Dict.get key2 p))) p ) ) model

This basicly 'avoids' the javascript pattern of 
`  update: while (true) { var _p0 = act;  ... continue; ...}` 
It seems to be something with tail recursion.

**Comment by jvoigtlaender** [comment]
Yes, it's a compiler bug with tail recursion. Your original `update` function is currently translated to the following JavaScript:
```js
   var update = F2(function (act,model) {
      update: while (true) {
         var _p0 = act;
         if (_p0.ctor === "ActOnFirstLevel") {
               var _p1 = _p0._0;
               return A3($Dict.insert,
               _p1,
               _p0._1(A2($Maybe.withDefault,
               $Dict.empty,
               A2($Dict.get,_p1,model))),
               model);
            } else {
               var _p2 = _p0._1;
               var _v1 = A2(ActOnFirstLevel,
               _p0._0,
               function (p) {
                  return A3($Dict.insert,
                  _p2,
                  _p0._2(A2($Maybe.withDefault,
                  "DefaultData",
                  A2($Dict.get,_p2,p))),
                  p);
               }),
               _v2 = model;
               act = _v1;
               model = _v2;
               continue update;
            }
      }
   });
```
The problem is the line
```js
               act = _v1;
```
Correct there would be
```js
               _p0 = _v1;
```
So somehow, a renaming is not carried through enough. I think @JoeyEremondi wrote the tail recursion optimization pass. Maybe he can quickly identify where the compiler is forgetting a rename.

**Comment by JoeyEremondi** [comment]
I'll take a look, but Evan modified my TR code quite heavily, so I'll have
to take some time to figure out what's going on.

**Comment by jvoigtlaender** [comment]
Oh, actually, I was just going to write I am not so sure of my diagnosis anymore.

**Comment by jvoigtlaender** [comment]
At least the `act = _v1;` is not obviously wrong as I claimed. After all, `_p0` will be set to `act` right at the beginning of the next loop iteration.

**Comment by jvoigtlaender** [comment]
How about this: Shouldn't there be a `var` before `_v2 = model;`? Else, where is `_v2` declared?

**Comment by AZon8** [comment]
There is a var before `_v2 = model` . 
it is declared in multi assignment mode. 

       var _v1 = A2(.....),
       _v2 = model

**Comment by AZon8** [comment]
For the time being , the following works for me 


      ActOnSecondLevel  key1 key2 updater -> let 
                    up = identity  updater
                in 
      update (ActOnFirstLevel key1 (\p ->  Dict.insert key2 (up (Maybe.withDefault "DefaultData" (Dict.get key2 p))) p ) ) model

**Comment by jvoigtlaender** [comment]
@AZon8, about the `var`, yes, I was blind to that.

But *now* I think I know what the issue is. And it seems rather serious/deep. Here goes:

The function
```js
               function (p) {
                  return A3($Dict.insert,
                  _p2,
                  _p0._2(A2($Maybe.withDefault,
                  "DefaultData",
                  A2($Dict.get,_p2,p))),
                  p);
```
refers `_p0`. But it is not closed over this, so as to capture the value of `_p0` *current at the point when that function is created*. Instead, it refers to the variable `_p0` in the overall `while` block. So when the time comes that this `function (p) { ... }` is applied, the content of variable `_p0` will have changed. And in fact it will not anymore be the case that its `_p0._2` contains a/the function one wants to apply. Bang!

**Comment by jvoigtlaender** [comment]
BTW, in a slightly different constellation, the same problem could occur with `_p2`.

**Comment by jvoigtlaender** [comment]
@evancz, this code generation bug is somewhat similar to https://github.com/elm-lang/elm-compiler/issues/1223.

**Comment by jvoigtlaender** [comment]
Here's a simpler example showing, I think, the same problem:
```elm
import Html exposing (..)

f : List Int -> (Int -> Int) -> Int
f xs g = case xs of
   [] -> g 0
   y::ys -> f ys (\_ -> y)

main = text (toString (f [1] identity))
```
That program should output "1". In fact it does if one destroys tail recursion by replacing the recursive call `f ys (\_ -> y)` with `0 + f ys (\_ -> y)`. But for the tail recursive form given, the compiled program blows up (try at http://elm-lang.org/try).

And I know how to fix it! It's really simple, actually. The JS code currently compiled from the above is:
```js
   var f = F2(function (xs,g) {
      f: while (true) {
         var _p0 = xs;
         if (_p0.ctor === "[]") {
               return g(0);
            } else {
               var _v1 = _p0._1,_v2 = function (_p1) {    return _p0._0;};
               xs = _v1;
               g = _v2;
               continue f;
            }
      }
   });
```
Replacing the line
```js
         var _p0 = xs;
```
by
```js
         let _p0 = xs;
```
fixes the program.

In general, further occurrences of `var` would have to be replaced by `let`. See, for example, my comment further above about `_p2` in the other example.

**Comment by jvoigtlaender** [comment]
I think the issue title should be changed. This is not only about causing runtime errors. Actually, this bug can even make an Elm program (keep terminating but) *produce a wrong value*.

Consider this program:
```elm
import Html exposing (..)

f : List Int -> (Int -> Int) -> Int
f xs g = case xs of
   [] -> 0
   [_] -> g 0
   y::ys -> f ys (\_ -> y)

main = text (toString (f [1,2] identity))
```
On http://elm-lang.org/try, it produces output `2`. But that is mathematically wrong. (Hint: add `0 +` to the recursive call to see what the correct value would be.)

**Comment by jvoigtlaender** [comment]
Following a comment by @gyzerok on the mailing list, I tried whether `catch-try` could serve as a substitute for `let` here (the latter not existing in Safari, for example). The answer appears to be yes. For the latest example above, where Elm currently computes the wrong result, rewriting
```js
   var f = F2(function (xs,g) {
      f: while (true) {
         var _p0 = xs;
         if (_p0.ctor === "[]") {
               return 0;
            } else {
               if (_p0._1.ctor === "[]") {
                     return g(0);
                  } else {
                     var _v1 = _p0._1,_v2 = function (_p1) {    return _p0._0;};
                     xs = _v1;
                     g = _v2;
                     continue f;
                  }
            }
      }
   });
```
to
```js
   var f = F2(function (xs,g) {
      f: while (true) {
         try { throw xs } catch (_p0) {
            if (_p0.ctor === "[]") {
                  return 0;
               } else {
                  if (_p0._1.ctor === "[]") {
                        return g(0);
                     } else {
                        var _v1 = _p0._1,_v2 = function (_p1) {    return _p0._0;};
                        xs = _v1;
                        g = _v2;
                        continue f;
                     }
               }
         }
      }
   });
```
fixes the issue.

**Comment by jwmerrill** [comment]
try/catch prevents the function that contains it from being optimized in v8 (and maybe other engines) [ref](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers). I would avoid using it just to simulate block scoping.

I'm not sure what the best solution is there, but seems like more aggressive variable name mangling (to avoid collisions in cases like this) might be an easy fix.

**Comment by jvoigtlaender** [comment]
Yeah, Evan already indicated on the mailing list that he is going to fix this by other means than let (or, I guess, an equivalent of it).

About easy fix, no, I don't see how name mangling might help here. Note that such loops could be iterated through an arbitrary number of times. And the accumulation of values could be such that values from each iteration are needed at the end. So in general, for a fix there must be allocation of new variables going on (as let would do; IIFEs might also be an option). It's not enough to just rename existing variables. At least I am confident that I can massage my example above to twart any such attempt.

**Comment by evancz** [comment]
I'm going to close this issue because it is hard to follow. @jvoigtlaender, can you turn your minimal repro into a new issue that explains exactly what's going on?
